zerver/models/realms.py

```diff
@@ -870,12 +870,26 @@ def upload_quota_bytes(self) -> Optional[int]:
         lambda realm: get_realm_used_upload_space_cache_key(realm.id), timeout=3600 * 24 * 7
     )
     def currently_used_upload_space_bytes(realm) -> int:  # noqa: N805
+        from analytics.models import RealmCount, installation_epoch
         from zerver.models import Attachment
 
-        used_space = Attachment.objects.filter(realm=realm).aggregate(Sum("size"))["size__sum"]
-        if used_space is None:
-            return 0
-        return used_space
+        try:
+            latest_count_stat = RealmCount.objects.filter(
+                realm=realm, property="used_upload_space::day"
+            ).latest("end_time")
+            last_recorded_used_space = latest_count_stat.value
+            last_recorded_date = latest_count_stat.end_time
+        except RealmCount.DoesNotExist:
+            last_recorded_used_space = 0
+            last_recorded_date = installation_epoch()
+
+        newly_used_space = Attachment.objects.filter(
+            realm=realm, create_time__gte=last_recorded_date
+        ).aggregate(Sum("size"))["size__sum"]
```

Comment:
So the strategy is to grab the latest CountStat we have about the usage, and only walk the most recent bit of the `Attachment` table that hasn't gone into the CountStat records yet.

Reply:
Note: I didn't change the fact that this function is cached:
```
    @cache_with_key(
        lambda realm: get_realm_used_upload_space_cache_key(realm.id), timeout=3600 * 24 * 7
    )
```
since this still lets us avoid some unnecessary queries, right? And when the cache gets invalidated by a file upload, we have the benefit of not calculating the sum through the whole table.

Reply:
Yeah this strategy is solid, we just need to add a database index on `(realm, create_time)` on `Attachment` to make the fixup query fast.

Reply:
Added an index in a migration